miniapps/solvers/plor_solvers.cpp

```diff
@@ -0,0 +1,220 @@
+// Copyright (c) 2010-2021, Lawrence Livermore National Security, LLC. Produced
+// at the Lawrence Livermore National Laboratory. All Rights reserved. See files
+// LICENSE and NOTICE for details. LLNL-CODE-806117.
+//
+// This file is part of the MFEM library. For more information and source code
+// availability visit https://mfem.org.
+//
+// MFEM is free software; you can redistribute it and/or modify it under the
+// terms of the BSD-3 license. We welcome feedback and contributions, see file
+// CONTRIBUTING.md for details.
+//
+//                 ------------------------------------------
+//                 Parallel Low-Order Refined Solvers Miniapp
+//                 ------------------------------------------
+//
+// This miniapp illustrates the use of low-order refined preconditioners for
+// finite element problems defined using H1, H(curl), H(div), or L2 finite
+// element spaces. The following problems are solved, depending on the chosen
+// finite element space:
+//
+// H1 and L2: definite Helmholtz problem, u - Delta u = f
+//            (in L2 discretized using the symmetric interior penalty DG method)
+//
+// H(curl):   definite Maxwell problem, u + curl curl u = f
+//
+// H(div):    grad-div problem, u - grad(div u) = f
+//
+// In each case, the high-order finite element problem is preconditioned using a
+// low-order finite element discretization defined on a Gauss-Lobatto refined
+// mesh. The low-order problem is solved using hypre's AMG preconditioners:
+// BoomerAMG is used for H1 and L2 problems, AMS is used for H(curl) and 2D
+// H(div) problems, and ADS is used for 3D H(div) problems.
+//
+// For vector finite element spaces, the special "Integrated" basis type is used
+// to obtain spectral equivalence between the high-order and low-order refined
+// discretizations. This basis is defined in reference [1] and spectral
+// equivalence is shown in [2]:
+//
+// [1]. M. Gerritsma. Edge functions for spectral element methods. Spectral and
+//      High Order Methods for Partial Differential Equations (2010).
+// [2]. C. Dohrmann. Spectral equivalence properties of higher-order tensor
+//      product finite elements and applications to preconditioning. (2021)
+//
+// The action of the high-order operator is computed using MFEM's partial
+// assembly/matrix-free algorithms (except in the case of L2, which remains
+// future work).
+//
+// Sample runs:
+//
+//    mpirun -np 4 plor_solvers -m ../../data/fichera.mesh -fe h
+//    mpirun -np 4 plor_solvers -m ../../data/fichera.mesh -fe n
+//    mpirun -np 4 plor_solvers -m ../../data/fichera.mesh -fe r
```

Comment:
Since the exact solution is for the Maxwell problem, the H(div) test does not converge as the order is increased. This could be confusing to users. We could add an exact solution for the H(div) case, or just not output the error in that case.

Reply:
There should be a different forcing term in the case of H(div). Are you really not seeing convergence? What commands are you running and what errors do you get?

Reply:
I see now that there is a different forcing term for H(div), but I don't see convergence for this test. Also, the solution plot looks weird on some parts of the boundary.
mpirun -np 4 plor_solvers -m ../../data/fichera.mesh -fe r
-o 3
L2 error: 0.0964843 
-o 4
L2 error: 0.0929569

Reply:
Good catch, there was a bug in `RT_HexahedronElement::ProjectIntegrated`. Can you check if it's working properly for you now?

Reply:
Yes, it works now.